### ISSUE TITLE
Generates duplicate mount points on Mac OS almost every reboot

### ISSUE THREAD

**Issue by DPattee** [minio/minio#8521]
## Expected Behavior
Expect minio server to start up and if the volume indicated is not available to terminate.

## Current Behavior
If minio server starts up just barely too fast and the target volume passed on the command line is not yet mounted, it will generate a mount point for that volume. Then when the volume does get properly mounted, the OS will see a mount point already exists with the correct name and will automatically generate a new mount point with a number appended, i.e. "TargetName 1".
This then causes minio to not function properly, and also causes all other services on the server to fail because the data they need is no longer on the path expected.

## Possible Solution
Check to see if the volume is already mounted. If not, fail the minio server launch.

## Steps to Reproduce (for bugs)
1. Mac OS 10.14.6
2. External hard drive (named 'Foo' for this example)
3. Folder at root of external drive for minio to store (named 'MinioDataTest' for this example)
4. Install minio server via homebrew (version 2019-09-18T21:55:05Z right now, but problem has existed across multiple updates
5. Launch daemon that will start up the minio server and keep it running if it crashes. Using "/usr/local/bin/minio server /Volumes/Foo/miniodatatest" for this example
6. Configure minio, send some files to it, etc all fine
7. Have weekly scheduled restart configured on the Mac server
8. Wait until the restart happens, come back and discover that many of your services didn't start
9. from terminal do an "ls /Volumes" and note that you have a small read only "/Volumes/Foo" that only contains a non-functional miniodatatest folder and a large read/write volume called "Volumes/Foo 1" that has all teh real folders and real data
10. Force delete the fake volume (i.e. rm -r), unmount the real volume, disable the launchdeamon that starts minio, reboot and let everything mount up properly

## Context
I can't use minio server on my Mac server because it fails almost every time the server reboots, and that takes down other services when minio mucks with the volume mounting process.

No other service causes this problem (i.e. elastic search is on that same volume, and it properly starts up under normal conditions)

## Regression
No, this has happened for months not sure what minio version I started using though.

## Your Environment

* Version used (`minio version`): 2019-09-18T21:55:05Z
* Environment name and version (e.g. nginx 1.9.1):Mac OS 10.14.6
* Server type and version:
* Operating System and version (`uname -a`):
* Link to your project:

**Comment by kannappanr** [comment]
@DPattee Will take a look and get back to you.

**Comment by BigUstad** [comment]
I tried to replicate this issue on a 10.14.6 machine but ran into issues. After researching online, I found that the best way to deal with external drives & processes that use the external drives in Mac is to use Automator.
These steps resolved it for me. They're a suggestion, may not be the standard steps.
1. With Automator (v2.9 with the machine). New Workflow.
2. Action -> Run AppleScript.
    2a. In the script editor: do shell script "/usr/bin/hdiutil mount <The_Drive>"
3. Compose a shell script (.sh) that will start minio with the appropriate exported variables. If possible, redirect output to files to refer later. 
Action -> Run AppleScript. Another one.
   3a. In the script editor: do shell script "<Path to above shell-script>"
4. Save as 'Application'.
5. System Preferences -> Users & Groups -> Login Items. Add the created 'Application'. Hide, if necessary.

Let me know if it works.

**Comment by DPattee** [comment]
That does seem like a temporary work around possible.

But,
1. most folks wouldn't rely on AppleScript in production
2. it would change the startup process of this service compared to most other services
3. it would tie the service startup to an auto login interactive user instead of system startup root (i.e. switching from a launchdaemon to a launchagent). this might tie the lifetime of the minio server to the lifetime of that user session too, not sure. 

Those instruction don't really have any indication on why you'd use Automator & AppleScript anyway, since all it is doing is running two shell commands. (hdiutil and the normal minio startup command). You could execute the hditutil script through the same command line that starts minio up in the first place... Besides, by the time an interactive user session could complete its login process, the volume would have been available for a long time already. 

A problem with hdiutil though is 
4. hdiutil mount relies on physical drive identifier not volume name, so I'd expect the script could fail if one of multiple esata arrays happen to be detected slightly faster than another on any particular boot, or if one (unrelated to minio) completely failed, or if you created a new partition on an array that had a partition related to minio.

And with a two-step launch process, 
5. I'll need to investigate if this would also make launchd's 'keep alive' functionality difficult to use since it either wouldn't be the process actually launching the minio process anymore (if you put the two commands in a shell script) or launchd would be running a command with two separate processes/exit codes and I just don't know what its keep alive behavior is in that case.

**Comment by BigUstad** [comment]
Thank you DPattee. As I did mention, my steps weren't standard and I agree with your reservations about using them in a production environment. My apologies for the confusion. However in my trials to reproduce the bug, I did make sure that the drive I was testing with was loaded at startup by adding a plist entry in /Library/LaunchDaemons and using a shell script to explicitly run the command to load the drive. However that one (load drive script) did have a disabled 'keep alive' and had 'launch only once' key enabled.  So, I agree with your comments about the two-step launch process.
Also, in my tests since the drive access wasn't enabled, minio exited gracefully and I didn't see a duplicate mountpoint created. I tested with the minio version available in brew: minio: stable 20191012013957. Can you please share more information about the drive and how it is available/loaded etc so we can replicate the issue?

**Comment by harshavardhana** [comment]
AFAICS the issue here seems like out of scope for MinIO - how you mount things in order on macOS is out of our scope.  MinIO all it needs is a path where either its a disk or a directory, MinIO cannot automatically do anything on your system unless you made your init scripts wrong.

We will close this issue out since it's not in our scope to be addressing your daemon launch concerns.  Thanks, @BigUstad for working on reproducing this - this is not in our scope.
